src/poprox_recommender/pipeline.py

```diff
@@ -0,0 +1,56 @@
+from copy import deepcopy
+from dataclasses import dataclass
+from typing import Callable
+
+from poprox_concepts import ArticleSet, InterestProfile
+
+
+@dataclass
+class ComponentSpec:
+    component: Callable
+    inputs: list[str]
+    output: str
+
+
+class RecommendationPipeline:
+    def __init__(self, name):
+        self.name = name
+        self.components = []
+
+    def add(self, component: Callable, inputs: list[str], output: str):
```

Comment:
Yeah, I like this. Let's try it!

Reply:
One wobble: going from `output` to `name` makes it less clear what's happening and how to name things. For example, the existing pipeline definition for our default recommender starts to look wrong, because it seems like multiple components are getting the same name:
```
pipeline = RecommendationPipeline(name=diversify)
pipeline.add(article_embedder, name="candidate", inputs=["candidate"])
pipeline.add(article_embedder, name="clicked", inputs=["clicked"])
pipeline.add(user_embedder, name="profile", inputs=["clicked", "profile"])
pipeline.add(article_scorer, name="candidate", inputs=["candidate", "profile"])
pipeline.add(diversifier, name="recs", inputs=["candidate", "profile"])
```

But giving the outputs different names makes it confusing whether the names refer to the components or the outputs, what the types of the outputs are, and what they included. Does `user_embed` output the whole interest profile or just the user embedding?
```
pipeline = RecommendationPipeline(name=diversify)
pipeline.add(article_embedder, name="cand_embed", inputs=["candidate"])
pipeline.add(article_embedder, name="click_embed", inputs=["clicked"])
pipeline.add(user_embedder, name="user_embed", inputs=["click_embed", "profile"])
pipeline.add(article_scorer, name="scored_articles", inputs=["cand_embed", "user_embed"])
pipeline.add(diversifier, name="recs", inputs=["scored_articles", "user_embed"])
```

Reply:
Another wobble: the API now depends on how individual components name their parameters, which may not be consistent across components (especially as experimenters start writing their own.)

```
pipeline = RecommendationPipeline(name=diversify)
pipeline.add(article_embedder, output="candidate", article_set="candidate")
pipeline.add(article_embedder, output="clicked", article_set="clicked")
pipeline.add(user_embedder, output="profile", clicked_articles="clicked", interest_profile="profile")
pipeline.add(article_scorer, output="candidate", candidate_articles="candidate", interest_profile="profile")
pipeline.add(diversifier, output="recs", candidate_articles="candidate", interest_profile="profile")
```

Reply:
First wobble: I'm fine with keeping `output` and making it a keyword parameter.

Second wobble: I see that as a bit of a feature, actually:

* It may encourage consistency
* It raises a failure that results in doc-consultation (mismatched name, so one looks at the docs, hopefully)
* IMO it's a worthwhile tradeoff to detect / prevent the more subtle and harder-to-detect error where two components do take the same inputs, but in a different order

Reply:
I think we can address the subtle error with better type checking of the arguments against the signature inside the pipeline code, so I'm less worried about that than you are. I'm also skeptical that people will actually read documentation even if something looks weird and/or doesn't work the way they expect. 😅 

Could we meet in the middle with automated type checking raising very verbose errors that border on documentation?